interactive todo app required files and folders added

BikramRawat opened the pull request:
react/week2

BikramRawat committed:
interactive todo app required files and folders added

BikramRawat committed:
unnecessary files deleted

BikramRawat committed:
Counter & other components updated with new features1

abdualrhman commented:
Hello Bikram, 

Good job with the homework. The interactive part also looks great!

abdualrhman commented:
Unfortunately, after some digging, I found a little bug in the program:

When an item in the todo list is marked as `done`, then the first element of the list will always be marked as `done`.
I left a comment in the code to how to fix it.

https://user-images.githubusercontent.com/28038265/133941750-4781053c-dbbc-4438-81d1-e73422812bbd.mov

abdualrhman commented on react/week2/interactive-todo-list/src/components/CreateTodo.jsx:
```diff
@@ -0,0 +1,83 @@
+
+  const handleChange = (event) => {
+    let { value } = event.target;
+    let obj = {};
+    obj["title"] = value;
+    obj["done"] = false;
+    setTodo(obj);
```
The problem is because you use indexing to identify the todo which is changing when the list is updated. 
Therefor we have to add `id` to each todo
```suggestion
    let obj = {};
    obj["title"] = value;
    obj["done"] = false;
    obj["id"]= Date.now()
    setTodo(obj);
```

the `Date.now()` is not best practice to get a unique id, but it's a quick fix for now

BikramRawat replied:
ok , thanks a lot for the suggestion, so we need to avoid using index for identifying the todo.

abdualrhman commented on react/week2/interactive-todo-list/src/components/CreateTodo.jsx:
```diff
@@ -0,0 +1,83 @@
+  };
+
+  const completeTodo = (i) => {
+    const updatedTodoArray = [...todoArr];
+    if (updatedTodoArray[i]["done"] !== true) {
+      updatedTodoArray[i]["done"] = true;
+      setTodoArr(updatedTodoArray);
+      swal("Good Job !", "Todo completed", "success");
+    }
```
We will now use the `id` to identify the updated todo:

```suggestion
    const updatedTodoArray = [...todoArr];
    const index = updatedTodoArray.findIndex((todo) => todo.id === i);
    if (updatedTodoArray[index]["done"] !== true) {
      updatedTodoArray[index]["done"] = true;
      setTodoArr(updatedTodoArray);
      swal("Good Job !", "Todo completed", "success");
    }
```

BikramRawat replied:
yes i got the point, thanks a lot

abdualrhman commented on react/week2/interactive-todo-list/src/components/TodoList.jsx:
```diff
@@ -0,0 +1,29 @@
+    <div>
+      <ul>
+        {todoArr && todoArr.length > 0 ? (
+          todoArr.map((el, i) => (
+            <li key={i}>
+              <div className={el["done"] ? "strike-through" : null}>
+                {el.title}
+              </div>
+              <div>
+                <input type="checkbox" onClick={() => completeTodo(i)} />
```
Here, we pass the `id` for both the `key` and `completeTodo`

```suggestion
          todoArr.map((el, i) => (
            <li key={el.id}>
              <div className={el["done"] ? "strike-through" : null}>
                {el.title}
              </div>
              <div>
                <input type="checkbox" onClick={() => completeTodo(el.id)} />
```

BikramRawat committed:
used id instead of index to identify the todo

Co-authored-by: Abdulrahman <abdulrahman.ramdan@outlook.com>

BikramRawat committed:
used id to identify the updated todo

Co-authored-by: Abdulrahman <abdulrahman.ramdan@outlook.com>

BikramRawat committed:
passed id to both key and the completeTodo

indexing is bad practice to identify the todos in the array so now converted to id for now id is taking Date.now()

Co-authored-by: Abdulrahman <abdulrahman.ramdan@outlook.com>

BikramRawat committed:
added ids instead of index, added the case for uncheck the checkbox to convert done = true to done = false if we uncheck the previously checked todo
